bottomless/src/read.rs

```diff
@@ -55,3 +58,148 @@ impl BatchReader {
+        // ┌────────────┬──────────────────┬───────────────────────┬──────────────────┐
+        // │ Chunk size │ Number of chunks │ Amount for chunk size │ Cumulative total │
+        // ├────────────┼──────────────────┼───────────────────────┼──────────────────┤
+        // │ 5 MiB      │ 16               │ 80 MiB                │ 80 MiB           │
+        // ├────────────┼──────────────────┼───────────────────────┼──────────────────┤
+        // │ 10 MiB     │ 16               │ 160 MiB               │ 240 MiB          │
+        // ├────────────┼──────────────────┼───────────────────────┼──────────────────┤
+        // │ 25 MiB     │ 16               │ 400 MiB               │ 640 MiB          │
+        // ├────────────┼──────────────────┼───────────────────────┼──────────────────┤
+        // │ 50 MiB     │ 16               │ 800 MiB               │ 1.406 GiB        │
+        // ├────────────┼──────────────────┼───────────────────────┼──────────────────┤
+        // │ 100 MiB    │ 9935             │ 970.215 GiB           │ 971.621 GiB      │
+        // └────────────┴──────────────────┴───────────────────────┴──────────────────┘
+        //
+        // We can send up to 971 GiB in chunks, which is more than enough for the
+        // majority of use cases.
+        //
+        // The last chunk is reserved for the remaining of the `gzip_reader`
+        let chunk_size = chunk_sizes[((part / 16) as usize).min(chunk_sizes.len() - 1)];
+
+        let mut buffer = bytes::BytesMut::with_capacity(chunk_size);
+        loop {
+            let bytes_written = reader.read_buf(&mut buffer).await?;
+            // EOF or buffer is full
+            if bytes_written == 0 {
+                break;
+            }
+        }
+
+        // EOF
+        if buffer.is_empty() {
+            has_reached_eof = true;
+            break;
+        }
+
+        let part_out = client
+            .upload_part()
+            .bucket(bucket)
+            .key(key)
+            .upload_id(upload_id.clone())
+            .body(ByteStream::from(buffer.freeze()))
+            .part_number(part + 1)
+            .send()
+            .await?;
+
+        parts.push(
+            CompletedPart::builder()
+                .part_number(part + 1)
+                .e_tag(
+                    part_out
+                        .e_tag
+                        .ok_or_else(|| anyhow::anyhow!("e_tag missing from part upload"))?,
+                )
+                .build(),
+        );
+    }
+
+    // If the gzip stream has not reached EOF we need to send the last part to S3.
+    // Since we don't know the size of the stream and we can't be sure if it fits in
+    // memory, we save it into a file to allow streaming.
+    //
+    // This would only happen to databases that are around ~1 TiB.
+    if !has_reached_eof {
+        let mut last_chunk_path = std::env::temp_dir();
+        last_chunk_path.push(rand::random::<u32>().to_string());
```

Comment:
We already rely on `tempfile` in sqld, let's use it here as well? It also avoids the (however unlikely) conflicts on pseudorandom names

Reply:
Great! Didn't know we already depend on it 😄.

Reply:
done :)